tests/lib/rules/no-constant-condition.js

```diff
@@ -358,18 +371,6 @@ ruleTester.run("no-constant-condition", rule, {
             code: "if(''+[]) {}",
             errors: [{ messageId: "unexpected", type: "BinaryExpression" }]
         },
-        {
-            code: "if([a]==[a]) {}",
```

Comment:
This is a true error that we won't catch any more, but it is equivalent to a false positive (`if([a]==[b]) {}`) which used to be a false positive.

Reply:
Is `if([a]==[b]) {}` a false positive? Both sides are objects, so this should be comparison by reference, and thus always `false` (even when the elements are same)?

```js
[1]==[1] // false
```

Reply:
Ah, yes. Sorry, not sure what I was thinking.

That said, `[a]` cannot be considered "constant" for the purposes of `==` comparison unless we also know what we are comparing it to (and thus how it will be compared), and currently the rule does not try to determine that information.

```JavaScript
[1] == 1
true
[0] == 1
false
```

So, you're absolutely right that `if([a]==[b]) {}` is a constant condition, but I still think it's appropriate to treat `[a]` as a non-constant expression when comparing with `inBooleanPosition: false`.

Reply:
Will `no-constant-binary-expression` catch these cases?

* ArrayExpression === anything
* ArrayExpression !== anything
* ArrayExpression  == ArrayExpression
* ArrayExpression  != ArrayExpression

I'm asking because these look like realistic mistakes that will be no longer reported by `no-constant-condition`,  so it might be good to special-case them to retain that behavior. But if they will be reported by another rule, then there's no need for that.

Reply:
These would get caught by `no-constnat-binary-expression`:

* ArrayExpression === anything
* ArrayExpression !== anything

These would _not_:

* ArrayExpression == ArrayExpression
* ArrayExpression != ArrayExpression

However, I suspect (hope!) that most people would be using some form of the `eqeqeq` rule to avoid the many many pitfalls and foot-guns exposed by testing for `==` equality.

Reply:
> These would _not_:
> 
> * ArrayExpression == ArrayExpression
> * ArrayExpression != ArrayExpression

I've just left a question about this on the `no-constant-binary-expression` PR: https://github.com/eslint/eslint/pull/15296#discussion_r780776004